### ISSUE TITLE
chain/state store improvements: introduce a block cache and optimise existing 'has' cache

### ISSUE THREAD

**Issue by raulk** [filecoin-project/lotus#4752]
The hypothesis is that the VM and chain sync access hot values repeatedly, specifically HAMT nodes, block ranges, and actor states, over the course of various epochs.

Introducing a block cache would alleviate disk reads (sstable access), which are the largest hotspot right now.

Ideally we'd want this cache to be a CBOR object cache instead of a block cache, to avoid the cost of redeserialisation (CPU and heap allocs), but at this stage that's micro-optimisation, as there are other larger fish to fry.

Additionally, the current ARC cache takes up around 1.85% of CPU time. It's a small fraction, but moving to a different cache library like Ristretto may narrow this a sliver, while preserving or augmenting the hit ratio.

## Useful profile screenshots

<img width="319" alt="image" src="https://user-images.githubusercontent.com/1101242/98364681-23204f00-2029-11eb-9e34-e2d481e68c62.png">

<img width="1086" alt="image" src="https://user-images.githubusercontent.com/1101242/98365422-50b9c800-202a-11eb-8113-1be64eaa177b.png">

<img width="630" alt="image" src="https://user-images.githubusercontent.com/1101242/98365763-c887f280-202a-11eb-85fb-469cddba17d2.png">

**Comment by arajasek** [comment]
@raulk Is this something on your radar / being handled by you & your friends?

**Comment by jennijuju** [comment]
@vyzo Do you have any thoughts on this? Is this implemented or needs further implementation.

**Comment by raulk** [comment]
We implemented this in https://github.com/filecoin-project/lotus/pull/4771, but the effective performance gain wasn't as pronounced as we had anticipated. The real solution is the splitstore, which keeps the hot data (latest chain finalities) in a dedicated store/region. In the current implementation, the hot and cold stores are backed by badger, and they both rely on mmap/page caches to access data efficiently. Since the hot store is accessed continuously, its indices (SStables) and vlogs should be prioritised in the kernel page cache over the cold stores, thus acting like a cache that we expect to be good enough for the purposes of this issue.
